ArduCopter/mode_autorotate.cpp

```diff
@@ -111,145 +164,194 @@ void ModeAutorotate::run()
+            // Prevent running the initial entry functions again
+            _flags.entry_initial = false;
 
-            }
+        }
 
+        if (!hover_autorotation) {
             // Slowly change the target head speed until the target head speed matches the parameter defined value
-            if (g2.arot.get_rpm() > HEAD_SPEED_TARGET_RATIO*1.005f  ||  g2.arot.get_rpm() < HEAD_SPEED_TARGET_RATIO*0.995f) {
-                _target_head_speed -= _hs_decay*G_Dt;
+            float norm_rpm = g2.arot.get_rpm()/g2.arot.get_hs_set_point();
+            if (norm_rpm > HEAD_SPEED_TARGET_RATIO*1.005f  ||  norm_rpm < HEAD_SPEED_TARGET_RATIO*0.995f) {
+                _target_head_speed -= _hs_decay * last_loop_time_s;
             } else {
                 _target_head_speed = HEAD_SPEED_TARGET_RATIO;
             }
-
             // Set target head speed in head speed controller
             g2.arot.set_target_head_speed(_target_head_speed);
-
             // Run airspeed/attitude controller
-            g2.arot.set_dt(G_Dt);
             g2.arot.update_forward_speed_controller();
-
             // Retrieve pitch target
             _pitch_target = g2.arot.get_pitch();
-
             // Update controllers
-            _flags.bad_rpm = g2.arot.update_hs_glide_controller(G_Dt); //run head speed/ collective controller
+            _flags.bad_rpm = g2.arot.update_hs_glide_controller(); //run head speed/ collective controller
+        } else {
+            _pitch_target = 0.0f;
+            g2.arot.update_hover_autorotation_controller(); //run head speed/ collective controller
+            g2.arot.set_collective_minimum_drag(motors->get_coll_mid());
+            g2.arot.set_entry_sink_rate(curr_vel_z);
+            g2.arot.set_entry_alt(g2.arot.get_ground_distance());
+        }
+
+        break;
+    }
+
+    case Autorotation_Phase::SS_GLIDE: {
+        // Steady state glide functions to be run only once
+        if (_flags.ss_glide_initial == true) {
+
+            gcs().send_text(MAV_SEVERITY_INFO, "SS Glide Phase");
 
-            break;
+            // Set following trim low pass cut off frequency
+            g2.arot.set_col_cutoff_freq(g2.arot.get_col_glide_freq());
+
+            // Set desired forward speed target
+            g2.arot.set_desired_fwd_speed();
+
+            // Set target head speed in head speed controller
+            _target_head_speed = HEAD_SPEED_TARGET_RATIO;  //Ensure target hs is set to glide in case hs hasn't reached target for glide
+            g2.arot.set_target_head_speed(_target_head_speed);
+
+            // Prevent running the initial glide functions again
+            _flags.ss_glide_initial = false;
         }
 
-        case Autorotation_Phase::SS_GLIDE:
-        {
-            // Steady state glide functions to be run only once
-            if (_flags.ss_glide_initial == true) {
+        // Run airspeed/attitude controller
+        g2.arot.update_forward_speed_controller();
```

Comment:
There are lots of `g2.arot` calls, you could add a const reference to the mode so you can just do `arot`, would just make some of the lines a little shorter. and the code easier to review.